### ISSUE TITLE
Add "show interest" and "lose interest" hover delays to CSS

### ISSUE THREAD

**Issue by mfreed7** [w3c/csswg-drafts#9236]
Please see the Popover Hint and Hover explainer for more context:

  https://open-ui.org/components/popover-hint.research.explainer/ 

We're working on an addition to the Popover feature that allows authors to easily and declaratively get hover-triggering of popovers. (And to make sure all input modalities are covered, "hover" includes keyboard focus and also touch screen long-press.) The idea for now is to use the `popovertargetaction` attribute, with a new value called `interest` (see [this discussion](https://github.com/openui/open-ui/issues/767#issuecomment-1654177227) for context on the name) that enables this new mode of triggering. When `popovertargetaction=interest` on a button that points (via `popovertarget=foo`) to a popover, that popover will be triggered when one of these things happens:

- The user mouse-hovers the button *for a specified period of time*, or
- The user keyboard-focuses the button *for a specified period of time*, or
- The user long-presses the button *for a specified period of time*

The key phrase above, and the point of this issue, is that the delay should be developer-configurable. Leaving this delay up to the UA leads to issues and missed use cases. For example, the built-in tooltip that shows up as a result of the `title` attribute has one major, common complaint: the UA-imposed delay is too long. Other, more creative use cases of this feature would also require a customizable hover-delay. For that reason, I'd like to propose that this delay is set in CSS, via a new property:

```css
button[popovertargetaction=interest] {
  popover-show-delay: 0.5s;
}
```

There is another quite related, yet separate, set of behaviors that happens *after* the popover is shown. Typical "tooltip" use cases would also like the tooltip to be **hidden** after the tooltip **stops** being hovered for a period of time. To make this work, *both* the invoker (the button above) *and* the popover should be able to be hovered to maintain the popover in the open state. But once both of them have been de-hovered for a period of time, the popover can be closed. For this, I propose:

```css
[popover] {
  popover-hide-delay: 1s;
}
```

Note that in the above proposal, the "show" delay is applied to the button that triggers the popover, while the "hide" delay is applied to the popover. That's because those elements are where the action takes place. However, one could make the case that it'd be nice to instead put both properties on one element, and the `[popover]` would seem to be the natural place to put it. In that case, the above properties could also be combined into a single two-valued property (credit @una for the idea):

```css
[popover] {
  popover-delay: 0.5s 1s; /* first value = show delay, second value = hide delay */
}
```

Thoughts?

**Comment by mfreed7** [comment]
> Can this work at the level of matching selectors by inventing functional forms for them like `:focus(visible, 30ms), :hover(3s)` or adding arguments to forgiving selectors like `:where(0 0 0, 300ms / :target, :active)`?

I'd be worried about that approach, because then you'd need a CSS way to actually trigger the popover/behavior. I.e. you'd need `:focus(visible,30ms) { trigger-popover-now: yes; }`. So far (and as far as I can see it going), that really belongs in the HTML part of this feature, right? Also, I think the feature I've described doesn't have an equivalent pseudo state already. `:focus` gets the keyboard part, `:hover` gets the mouse hover part, nothing gets the long-press part, and (most importantly) nothing enforces that you need to use them all together or someone will be broken.

**Comment by tabatkins** [comment]
Yeah, this isn't really doable via Selectors, it's triggering a different behavior that's *not* CSS controlled. CSS doesn't make something a popover, that's a quality of it imposed by the DOM APIs (or the UA itself in some cases). So that's not a viable approach.

-------

Back to the OP, this seems reasonable.

I agree that it would make the most UX sense to set both the show and hide delays on the same element. I could see it being set on either the button *or* the popover, tho - as you say, both the popover *and* the invoker button need to be dehovered before the hide timer starts. Since you're setting the "interest" popover method on the button, and these are both effectively just applying some config to the "interest" method, I think it probably makes the most sense to put them both on the button, then?

One note is that we don't use CSS to configure/control non-CSS concepts much (at all?). Is it needed here, or should it be communicated in the DOM attribute? Why are the delays in CSS, but the indication that you want to use "interest" method is in the DOM? I presume the reason to put the delays into CSS is because they're unlikely to be different per-invoker, so you can reduce a lot of repetitive attribute values by making it a property, but would that similarly apply to the popover method?

(I think CSS *should* be used for non-CSS concepts a lot more, it's a rich and powerful language both in values and targeting, and authors are intimately familiar with it. I'm just wondering what the reasoning behind this particular slicing of responsibility is.)

**Comment by mfreed7** [comment]
> I agree that it would make the most UX sense to set both the show and hide delays on the same element. I could see it being set on either the button _or_ the popover, tho - as you say, both the popover _and_ the invoker button need to be dehovered before the hide timer starts. Since you're setting the "interest" popover method on the button, and these are both effectively just applying some config to the "interest" method, I think it probably makes the most sense to put them both on the button, then?

Interesting idea to put it on the button. I was initially thinking it made more sense to put them on the popover, since that's the thing that is showing and hiding. But I like the idea of adding both to the button; that also makes it possible to use the same popover invoked via different buttons that each have different delays. I'm not sure what the use case is, but it just feels more flexible which is good.


> One note is that we don't use CSS to configure/control non-CSS concepts much (at all?). Is it needed here, or should it be communicated in the DOM attribute? Why are the delays in CSS, but the indication that you want to use "interest" method is in the DOM? I presume the reason to put the delays into CSS is because they're unlikely to be different per-invoker, so you can reduce a lot of repetitive attribute values by making it a property, but would that similarly apply to the popover method?
> 
> (I think CSS _should_ be used for non-CSS concepts a lot more, it's a rich and powerful language both in values and targeting, and authors are intimately familiar with it. I'm just wondering what the reasoning behind this particular slicing of responsibility is.)

The reason I proposed using CSS for this is exactly what you said: it's just easier to do it there. I would bet that the overwelmingly-common thing is for *all* popover invokers on a page to use the same standard delays. So it's easier to do `button[interesttarget] {interest-delay: 0.5s 0.5s;}` once and not have to worry about adding attributes to all the buttons. The `interesttarget` attribute is in the DOM because (I think) it's somewhat semantic - it tells you these two elements are linked together via this connection. Also the a11y associations typically happen as a result of the DOM content and not as a result of anything from CSS.

**Comment by mfreed7** [comment]
Hello, I'd like to ping this issue as the result of a good discussion we just had in OpenUI related to the delays discussed on this issue:

https://github.com/openui/open-ui/issues/963#issuecomment-1856462744

There are a number of things discussed there, including the concept of a safe-area triangle. But related to this issue, a few interesting things were discussed. First, it does seem that folks like the idea of allowing CSS to control these delays, independently for the "show" and "hide" actions. In addition, we had a good discussion about specifying the delays themselves. As background, there are several issues with these delays:

- There likely need to be different delays between mouse, keyboard, and "other" input modalities. E.g. hovering with a mouse might need a longer delay than focusing an element with the keyboard. Things like eye tracking might need a separate way of thinking about delays entirely.
- There likely need to be different delays between hovering and de-hovering an element. You might want a longer de-hover delay, for example, so it's easy to get to the element before it goes away.
- It seems that most often, the choice of a particular delay should lie with the **user** rather than the **developer** since some folks prefer fast moving interfaces and others prefer more time to get around the screen.
- Having said the above, the developer also has a preference, since some interfaces are more naturally fast-moving (think of a game console or other fast-action UI) and others are less so (e.g. tooltips on Wikipedia).

Given the above constraints, the new idea is to specify the delays not as actual time values, but as **"slow"** or **"fast"**, etc. So:

```css
button[interesttarget] {
  interest-delay: [none|slow|normal|fast] [none|slow|normal|fast];
  /* interest-delay: interest-delay loseinterest-delay */
}
```

It would then be up to the UA to determine how to resolve "slow" into an actual delay value in seconds. The UA could also offer a user preference like "Minimal delays please" which would then reduce the delay value for all settings, or "I want more time" to increase them.

Thoughts? Suggestions?

**Comment by annevk** [comment]
Since we don't have a good handle on a cross-platform solution to "rich tooltips" WebKit doesn't think this is ready to move forward. See https://github.com/WebKit/standards-positions/issues/305#issuecomment-2231000262 for additional context.

**Comment by mfreed7** [comment]
> Since we don't have a good handle on a cross-platform solution to "rich tooltips" WebKit doesn't think this is ready to move forward. See [WebKit/standards-positions#305 (comment)](https://github.com/WebKit/standards-positions/issues/305#issuecomment-2231000262) for additional context.

Thanks. See also [my response](https://github.com/WebKit/standards-positions/issues/305#issuecomment-2287429935). This issue is about solving one small piece of the tooltip puzzle. You've raised other issues, such as touchscreen, on the WebKit standards position. I hope you can help us with solutions to the issues you've raised! But since this issue is unrelated, perhaps we can still discuss it?

**Comment by bramus** [comment]
One thing that came up in [this discussion on X _(then still Twitter)_](https://twitter.com/simevidas/status/1426442864976973826) was that a delay itself doesn’t always cut it.

Sometimes, you want the delay to only start counting from the moment the user has actively stopped over on the element instead of just a fly-by hover which just happens to be slow enough for the delay to expire.

/cc @simevidas

**Comment by mfreed7** [comment]
Thanks for the great discussion and ideas. I heard these general viewpoints (none were unanimous):

- CSS seems to be the right place to put this part of the feature.
- Many people seem to think it's better to allow explicit time values (e.g. `0.5s`) rather than `long`/`short` named values. It's still possible for UAs to override this for users, e.g. via a "make delays 10x longer" setting.
- There's a need for an infinite duration, e.g. for a tooltip that never goes away.
- There is a suggestion to have different delays when there's already a tooltip/menu showing. This might require a new pseudo-class like `:interest-target-showing` or something?
- If this feature allows explicit time durations, then we still need a way to specify different delays for different input modalities (hover, focus, touch, gaze). E.g. hover delays are typically longer than focus delays, which are typically also non-zero. I'm not sure how to do that.

Further ideas and discussion is welcome!

**Comment by kizu** [comment]
Comments promised in the call:

> @flackr: I wanted to make an argument against complete customisation. […] The argument for this is that the UA could provide a consistent experience across sites and the OS.

I think the web platform is rich enough to have vastly different types of apps and sites in it, where it might not be possible to normalize them to a specific set of values. Some apps might want to have snappier values, others — more relaxed. A work tool like a monitoring app will likely need a different set of values compared to something like a casual blog or a promo page. `short` and `long` could mean different things in the context of the corresponding design systems, and a pre-defined list of UA-only values couldn't cover all the use cases (but might be useful as the default consistent list).


> > @mfreed7: It's still possible for UAs to override this for users, e.g. via a "make delays 10x longer" setting.
> 
> @dholbert: I worry a little that with specifying exact time, developers would coordinate animations with delays which would look cool but would break with UA adjustments.

I like an idea of a global 10x longer setting. For cases where we'd need to synchronize things, we could introduce something similar to `forced-color-adjust` property, but for these delays/transitions.

**Comment by flackr** [comment]
My argument for not customizing this is that it empowers the UA to provide all of these good behaviors to all popover hints rather than requiring each developer to think about them, e.g.:
* Clear allowance to align with user / OS preferences,
* Can be based on a time after stopping movement: https://github.com/w3c/csswg-drafts/issues/9236#issuecomment-2288821778,
* Can eliminate or significantly reduce the delay for opening the n'th hint after a first one is open: based on discussion in https://github.com/w3c/csswg-drafts/issues/9236#issuecomment-2289285305,

I also feel like the use cases for instant reveal are fundamentally different than a "short" delay, and may be best represented by either a different named concept, e.g. hover.

> Some apps might want to have snappier values, others — more relaxed.

I worry that this sort of customization may not actually be providing a better user experience.

There were some questions during the call about touch - I think that just like click, contextmenu, :hover, and other "mouse-centric" properties where we already have a way to trigger with touch, we would do the same for this capability. My proposal which aligns with android behavior would be to show it after a short delay of pressing down - long enough to not be considered a click.

**Comment by kbabbitt** [comment]
Re: customizing delays vs not -- why not both? Provide at least an `auto` value that leaves it up to the UA to align to preferences, perhaps also `short` / `long` / etc. for variances where the author wants some variance but doesn't care about specific values, and also accept `<time>` for authors who have specific values in their design language - with the provision that UAs can still override those specific values using a forced colors like mechanism.

**Comment by kizu** [comment]
> Clear allowance to align with user / OS preferences,

A predefined list of keywords would be enough for most cases, while a custom duration will be an escape hatch for those cases where things are important for the authors. “Simple thing easy, complex — possible”.

> Can be based on a time after stopping movement

I don't think this contradicts a user-defined value. This could be a default behavior, with a potential additional option to opt-out of it (but I don't think it is necessary).

> Can eliminate or significantly reduce the delay for opening the n'th hint after a first one is open

The problem: there are cases where we'd like to have independent groups of tooltips. Two toolbars on two different sections would want to “share” one delay across items inside them, but not across sections.

> I worry that this sort of customization may not actually be providing a better user experience.

Relying solely on the OS/UA behavior will mean there will be no room for innovation, and as soon as someone will want to do something unusual and modify the experience a bit, they will have to opt out of the default behavior _completely_, rather than use an escape hatch. Which will lead to a much worse experience and potential footguns while trying to re-implement it with other means.

**Comment by flackr** [comment]
> Relying solely on the OS/UA behavior will mean there will be no room for innovation, and as soon as someone will want to do something unusual and modify the experience a bit, they will have to opt out of the default behavior completely, rather than use an escape hatch. Which will lead to a much worse experience and potential footguns while trying to re-implement it with other means.

So to be clear, I'm not saying that we shouldn't also allow some customizations, but I am concerned that if we try to design a bunch of strictly specified individual features we may make it harder for the UA to innovate while still honoring the developer specified values. My proposal is more that we determine what properties a common solution can have, and then add customizations as needed to meet the use cases (with clear examples) that are fundamentally different.

**Comment by flackr** [comment]
I built a demo to show what's possible using transitions on a custom property with style queries on that property:
https://codepen.io/flackr/pen/gONXNdV

**Comment by yisibl** [comment]
@mfreed7 I noticed that the `interest-target-delay` CSS property has been added to [Interest Invokers explainer](https://open-ui.org/components/interest-invokers.explainer/), so can it be used for `popover=hint`?

**Comment by mfreed7** [comment]
> [@mfreed7](https://github.com/mfreed7) I noticed that the `interest-target-delay` CSS property has been added to [Interest Invokers explainer](https://open-ui.org/components/interest-invokers.explainer/), so can it be used for `popover=hint`?

Yeah, I've taken the feedback from this issue (particularly the [live discussion](https://github.com/w3c/csswg-drafts/issues/9236#issuecomment-2289301749)) and incorporated it into the explainer, including the explicit delays and the `:has-interest` pseudo class. It's just a prototype/proposal at this point though, so it's not "done".

**Comment by mfreed7** [comment]
I'd like to bring this issue back around for discussion. Based on the [last discussion](https://github.com/w3c/csswg-drafts/issues/9236#issuecomment-2289301749), plus a bit of evolution of the [explainer](https://open-ui.org/components/interest-invokers.explainer/#mousekeyboard-delays), we might be able to get a resolution. In particular:

The current proposal is:

```
interest-target-show-delay = <time># 
interest-target-hide-delay = <time># 
interest-target-delay = <time>{1,2}
```

One open question is the naming, obviously. If we want to re-use these properties (perhaps without re-using the `interesttarget` API under the covers) for [`::tooltip`](https://github.com/w3c/csswg-drafts/issues/8930), then maybe we need a slightly more generic set of names?

Another open question is the default value. In the explainer I propose a default value of `0.5s` for reasons, but I could also see a natural default value of `0s`.

**Comment by una** [comment]
+1 to reducing this time significantly (and I would honestly prefer the browser default to be `0s`). Users could add their own delays as needed. I find this to be the more likely scenario.

**Comment by mfreed7** [comment]
Using `0s` as the default also makes sense to me. I did feel like `0.5s` felt a little too opinionated. The one counterpoint, which is easy to overcome with a little developer education, is that with a `0s` delay, if the hovercard isn't positioned directly next to the trigger, you can't mouse-move from trigger to hovercard without losing interest and hiding the hovercard. But developers can relatively easily just change the `-hide-` delay to fix this.

**Comment by tabatkins** [comment]
Even if the trigger and card *are* next to each other, mouse movement that doesn't *perfectly* stay within the bounds of both of them will still close the card immediately if the delay is 0. This sort of interaction anti-pattern has been talked about a lot, and there are various ways to get around it (like creating a triangle- shaped invisible element extending from the target to the card, so all straight paths from anywhere in the target to anywhere in the card remain on a child that triggers :hover), but just having the "lose interest" delay be a small fraction of a second by default will do the job just fine. We generally shouldn't design an initial value that feels *broken* and requires author attention unless there's no way for us to predict what a non-broken value would be; in this case we can do so, so we should.

So I think the "lose interest" initial delay should still be non-zero (but maybe smaller than .5s, I dunno), and I don't have too much of an opinion on whether the "show interest" initial value should be zero or non-zero. 

(I do think the initial values should be `normal` or `auto`, tho, rather than a fixed number, to allow for platform divergence or us adjusting things as we go.)

**Comment by bkardell** [comment]
> like creating a triangle- shaped invisible element extending from the target to the card, so all straight paths from anywhere in the target to anywhere in the card remain on a child that triggers :hover

I'm curious if you are suggesting this is a thing that developers should do in order to use this feature?

**Comment by tabatkins** [comment]
Oh, absolutely not. It's a workaround for making the "zero delay" behavior still work reasonably with common user behavior. If we just have a reasonable non-zero delay, that's not needed. (And the delay is *better* than the triangle - if the user just stops and leaves their mouse in the triangle, you actually want it to still count as losing interest.)

We should just make the default lose-interest delay non-zero, to avoid that sort of bad behavior by default.

**Comment by una** [comment]
I would be okay with keeping a lost-interest delay and having a gain-interest delay set at 0. That being said, I still think the lose-interest delay default should be quite fast. Doing some testing, `0.05s` *feels* the best. I wouldn't go slower than `0.1s` for the default values. If users need a longer delay, they can make that adjustment.

**Comment by bramus** [comment]
+1 on not using a non-zero lose-interest delay for reasons outlined by @tabatkins [above](https://github.com/w3c/csswg-drafts/issues/9236#issuecomment-2855465696). The initial values should be keywords indeed (`normal`?)

> ([#](https://github.com/w3c/csswg-drafts/issues/9236#issuecomment-2877997194)) I still think the lose-interest delay default should be quite fast. Doing some testing, `0.05s` _feels_ the best. I wouldn't go slower than `0.1s` for the default values.

In https://codepen.io/bramus/pen/raNLLor I landed on these values when hovering with the mouse:

```css
interest-target-show-delay: 0.25s;
interest-target-hide-delay: 0.10s;
```

What I also found was that I wanted much much shorter delays when tabbing into the element _(versus hovering it)_ but that is food for a follow-up issue.

**Comment by tabatkins** [comment]
Spec text checked in. Pretty trivial definition-wise, so I don't think it needs additional WG review, but @mfreed7 please verify that this works for your HTML PR.
